quinn-proto/src/connection.rs

```diff
@@ -1767,13 +1764,13 @@ impl Connection {
         }
 
         self.app_closed = true;
-        self.state = Some(match self.state.take().unwrap() {
+        self.state = match &self.state {
             State::Handshake(_) => State::Closed(state::Closed { reason }),
             State::Established => State::Closed(state::Closed { reason }),
-            State::Closed(x) => State::Closed(x),
+            State::Closed(x) => State::Closed(x.clone()),
             State::Draining => State::Draining,
             State::Drained => unreachable!(),
-        });
+        };
```

Comment:
This could be made more concise and clone-free by moving the assignment into the `Handshake`/`Established` cases and removing the no-op cases entirely.

Reply:
Yup, good idea.